### DISCUSSION TITLE
Allow tweaking low-level `RigidBodyBuilder`

### DISCUSSION THREAD

**Discussion by MOZGIII** [jcornaz/heron#163] [Ideas]
I'd like a capability to tweak `RigidBodyBuilder`'s parameters at rigid body spawn time in an arbitrary manner.

My concrete interest for now is turning on `ccd_enabled` - but something else might be required later (I can imagine `sleeping` and related stuff).

**Comment by jcornaz** [comment]
Hi,

For each of theses needs, I would rather improve heron to add a dedicated component to the API (but only when actually needed). So for your current need we could feature a `ContinuousCollisionDetection` component. And we may add a `Sleeping` component when the need will arise for someone.

I understand it means that heron's is more limited, as one has to wait until heron adds support for each feature.

But it is somewhat the point of this plugin. Users that want full power of rapier out-of-the box (at the cost of higher complexity) can use `bevy_rapier`, and the point of having `heron` as an alternative, is to provide a well-crafted data-oriented API (at the cost of being more limited).

**Reply by MOZGIII** [reply]
I just wanted to exercise the promise that heron allows low-level access when needed. In fact, if it doesn't I will switch, despite that, I tried both, and have just switched *from* `bevy_rapier` *to* `heron`, cause I like the design layout a bit better.

My use case is I need physics mostly for client-side cosmetics, all the gameplay systems won't be affected by the physics engine. So, I should be fine with limited integration, however, there *are* some things I would like to tweak that are considered low-level.

Given the above context, I think that `heron`, as long as the ability for low-level tweaking where needed is there, is a better choice for me at the moment. I am willing to take responsibility for misusing the available extension points if they are provided to me.

I'd be happy to build a proper integration of the tweaks I need into `heron`, but I just need to work on a different thing atm. But I'd rather keep using `heron`, than switch back and forth between it and `bevy_rapier`.

**Reply by jcornaz** [reply]
Yes of course I understand that.

The point of the readme you are referring to, is something that I would like to uphold. However, it is not the case atm.

But I have an Idea on how to make it true.

I cannot expand much on it right now, but I hope to be able to give more details, and  maybe even implement a solution today or tomorrow.

**Reply by jcornaz** [reply]
Mmh, actualy it *should* be already possible. Not well documented, nor nice, but possible.

You can create the rigid-body using `RigidBodyBuilder` and add the rigid body yourself to the `RigidBodySet` resource. Then you insert the handle returned by the RigidBodySet as a component alongs others heron components.

That way heron should recognise the rigid body and still manage it.

**Reply by MOZGIII** [reply]
I don't want to take over the complete creation of the `RigidBodyBuilder` though, just want to toggle a boolean.

This could be done by allowing some kind of hook: `fn myfn(rigid_body_builder: &mut RigidBodyBuilder)`.

If there were a resource that would allow such tweaking - I'd be happy.

That said - maybe the hook as described above is not the best API for the long term... So I'd do a couple more passes on this.

Maybe we can have a hook like this at a component level - so it's applied per archetype? Might be worth exploring...

**Reply by jcornaz** [reply]
If you want heron events to work when using this techinque, you have to write the bevy entity (as bits) in the rigid-body data.

**Reply by jcornaz** [reply]
> I don't want to take over the complete creation of the `RigidBodyBuilder` though, just want to toggle a boolean.
> 
> This could be done by allowing some kind of hook: `fn myfn(rigid_body_builder: &mut RigidBodyBuilder)`.
> 
> If there were a resource that would allow such tweaking - I'd be happy.
> 
> That said - maybe the hook as described above is not the best API for the long term... So I'd do a couple more passes on this.
> 
> Maybe we can have a hook like this at a component level - so it's applied per archetype? Might be worth exploring...

Yeah, maybe. But it would be a shame to spend too much time on desiging and implementing this instead of directly adding first-class support for a `ContinuousCollisionDetection` component.

**Comment by apastushenka** [comment]
> I don't want to take over the complete creation of the RigidBodyBuilder though, just want to toggle a boolean.

You can use code like this for adjusting a rigid body after creation:

```rust
use heron::rapier_plugin::rapier2d::dynamics::{RigidBodyHandle, RigidBodySet};

struct EnableCcd;

fn adjust_rigid_body(
    mut commands: Commands,
    mut bodies: ResMut<RigidBodySet>,
    handles: Query<(Entity, &RigidBodyHandle), With<EnableCcd>>,
) {
    for (entity, handle) in handles.iter() {
        if let Some(body) = bodies.get_mut(*handle) {
            body.enable_ccd(true);
            commands.entity(entity).remove::<EnableCcd>();

            println!("{:?}: enable ccd", entity);
        }
    }
}
```

But there is a pitfall (correct me if I'm wrong): according to functions `remove_invalids_after_component_changed` and `remove_invalids_after_components_removed` in [body.rs](https://github.com/jcornaz/heron/blob/main/rapier/src/body.rs) heron removes and creates again a rapier::RigidBody after `RigidBody`/`RotationConstraints`/`PhysicMaterial` component changed or removed, so your low-level settings of a rigid body will be lost in this case.
